include/aklite.h

```diff
@@ -0,0 +1,178 @@
+  Status status;
+  std::string description;
+};
+
+/**
+ * The response from an AkliteClient call to Install
+ */
+class InstallResult {
+ public:
+  enum class Status {
+    Ok,
+    NeedsCompletion,
+    Failed,
+  };
+  InstallResult(Status status, std::string description = "") : status(status), description(std::move(description)) {}
+  Status status;
+  std::string description;
+};
+
+/**
+ * AkliteClient provides an easy-to-use API for users wanting to customize
+ * the behavior of aktualiz-lite.
+ */
+class AkliteClient {
+ public:
+  /**
+   * Construct a client instance pulling in config files from the given
+   * locations. ex:
+   *
+   *   AkliteClient c(AkliteClient::CONFIG_DIRS)
+   *
+   * @param config_dirs The list of files/directories to parse sota toml from.
+   */
+  AkliteClient(const std::vector<boost::filesystem::path> &config_dirs);
+  /**
+   * Used for unit-testing purposes.
+   */
+  AkliteClient(std::shared_ptr<LiteClient> client) : client_(client) {}
+
+  /**
+   * Performs a "check-in" with the device-gateay. A check-in consists of:
+   *  1) ask device-gateway for a new root.json - normally a 404.
+   *  2) ask device-gateway for timestamp and snapshot metadata.
+   *  3) pull down a new targets.json if needed
+   */
+  CheckInResult CheckIn(std::string hwid = "") const;
+
+  /**
+   * Return the active aktualizr-lite configuration.
+   */
+  boost::property_tree::ptree GetConfig() const;
+
+  /**
+   * Report network and hwinfo to server. This only reports data if the
+   * information has changed locally.
+   */
+  void SendTelemetry() const;
+
+  /**
+   * Report the full aktualizr-lite active configuration to the
+   * device-gateway. Should be called once during start-up .
+   */
+  void ReportAkliteConfiguration() const;
+
+  /**
+   * Return the Target currently running on the system.
+   */
+  TufTarget GetCurrent() const;
+
+  /**
+   * Download and verify the content of the given target.
+   */
+  DownloadResult Download(const TufTarget &t, const std::string &correlation_id, std::string reason = "");
+
+  /**
+   * Install the Target
+   */
+  InstallResult Install(const TufTarget &t, const std::string &correlation_id);
+
+  /**
+   * Check if the Target has been installed but failed to boot. This would
+   * make this be considered a "rollback target" and one we shouldn't consider
+   * installing.
+   */
+  bool IsRollback(const TufTarget &t) const;
+
+  /**
+   * Default files/paths to search for sota toml when configuration client.
+   */
+  std::vector<boost::filesystem::path> CONFIG_DIRS = {"/usr/lib/sota/conf.d", "/var/sota/sota.toml",
+                                                      "/etc/sota/conf.d/"};
+
+  /**
+   * Return a random UUID to be used for a Download/Install operation
+   */
+  static std::string GenerateCorrelationId(const TufTarget &t);
+
+ private:
+  std::shared_ptr<LiteClient> client_;
```

Comment:
It doesn't look like something that should be part of the external API. Maybe this API should be pure interface (just set of pure virtual functions) or abstract class. So, this `std::shared_ptr<LiteClient> client_;` becomes a member of the interface/base-class subclass?

Reply:
I thought a bit about that. Starts to feel overkill. We wind up having to define everything twice. this makes a forward declaration to a private class. The contents of liteclient are opaque to the user. 

One unwritten decision I made while doing this - I'm not aiming for ABI compatibility. So, we could remove that in the future and clients would work properly after re-compiling.

Maybe once this API really settles down we should think about going more elaborate, but I kind of like the minimalism for now.

Reply:
Removing  the LiteClient from the interface makes it even more minimalistic :). 

 >We wind up having to define everything twice
Why?

Reply:
>Maybe once this API really settles down we should think about going more elaborate
Yes, it's not critical at this stage. We can elaborate it later. IMHO, any external API/interface should be pure virtual :), otherwise further development and support will be very hacky :).